### ISSUE TITLE
shaping problem in some fonts

### ISSUE THREAD

**Issue by realfinder** [harfbuzz/harfbuzz#2888]
hi

1. in Hacen Tehran
[Hacen Tehran_0.zip](https://github.com/harfbuzz/harfbuzz/files/6097262/Hacen.Tehran_0.zip)
`ريـكّــا، يومًا ما سيصبح شينوبي عظيم`
it will be
![image](https://user-images.githubusercontent.com/2512153/110242861-f5acc900-7f68-11eb-998b-887e9f6be912.png)
when it should be
![image](https://user-images.githubusercontent.com/2512153/110242913-33a9ed00-7f69-11eb-8908-dfc23547b98c.png)
it was same old one here https://github.com/libass/libass/issues/71#issuecomment-55250533 (still no fix for years)

2. in Hacen Typographer Bold mod
[Hacen Typographer Bold mod.zip](https://github.com/harfbuzz/harfbuzz/files/6097275/Hacen.Typographer.Bold.mod.zip)
`كَرم`
it will be as
![image](https://user-images.githubusercontent.com/2512153/110243072-e67a4b00-7f69-11eb-9714-1c02121c7e5e.png)

thanks

**Comment by khaledhosny** [comment]
Please open a separate bug report for each issue next time.

1. Seems to be another instance of #2647. The font has mark ligature lookup with “ignore base” flag set, but an empty `GDEF` table.
2. The font has `mset` feature that maps `fatha` glyph to `fathaisolated` glyph, and the later has a +ve advance width but not assigned mark class `GDEF` table, so HarfBuzz does not zero its width. Uniscribe seems to not apply the `mset` feature.
3. Similar to 2., HarfBuzz is applying the `mset` feature, but Uniscribe is not.

**Comment by realfinder** [comment]
sorry, I do not have enough knowledge of these font things, so it's something need to be fix in libass not HarfBuzz? or HarfBuzz? or both?

from quick search Uniscribe seems in windows only, if it fixed in Uniscribe then other OSes will have no fix?

**Comment by khaledhosny** [comment]
> 2. The font has `mset` feature that maps `fatha` glyph to `fathaisolated` glyph, and the later has a +ve advance width but not assigned mark class `GDEF` table, so HarfBuzz does not zero its width. Uniscribe seems to not apply the `mset` feature.
> 
>3. Similar to 2., HarfBuzz is applying the `mset` feature, but Uniscribe is not.

I retested again and I take this back, `mset` feature is applied and I get identical result in from 3. on Windows Notepad, so I think the font is working as design:
![Capture](https://user-images.githubusercontent.com/93914/110352204-522de800-803e-11eb-8579-11bad7f4242e.PNG)

For 1. and 2., it seems to be an issue of zeroing the advance width of mark glyphs when they are not assigned a glyph class in `GDEF` table.

**Comment by realfinder** [comment]
> 
> 
> > 1. The font has `mset` feature that maps `fatha` glyph to `fathaisolated` glyph, and the later has a +ve advance width but not assigned mark class `GDEF` table, so HarfBuzz does not zero its width. Uniscribe seems to not apply the `mset` feature.
> > 2. Similar to 2., HarfBuzz is applying the `mset` feature, but Uniscribe is not.
> 
> I retested again and I take this back, `mset` feature is applied and I get identical result in from 3. on Windows Notepad, so I think the font is working as design:
> ![Capture](https://user-images.githubusercontent.com/93914/110352204-522de800-803e-11eb-8579-11bad7f4242e.PNG)

so that mean the vsfilter did it wrong here? since https://user-images.githubusercontent.com/2512153/110243478-9ac8a100-7f6b-11eb-8dad-6d8e6339fadd.png was made with it

**Comment by khaledhosny** [comment]
I don’t know how to test vsfilter, but I downloaded https://sourceforge.net/projects/mpc-hc/ and got the same result:
![Capture](https://user-images.githubusercontent.com/93914/110372288-ca080c80-8056-11eb-97bc-07cdb3e8c6c1.PNG)

**Comment by realfinder** [comment]
> 
> 
> I don’t know how to test vsfilter, but I downloaded https://sourceforge.net/projects/mpc-hc/ and got the same result:
> ![Capture](https://user-images.githubusercontent.com/93914/110372288-ca080c80-8056-11eb-97bc-07cdb3e8c6c1.PNG)

true, seems it was some windows compatibility problem same as the one here https://github.com/libass/libass/issues/374#issuecomment-792337087, I was on win7 with UK system back then, now win10 with US one and I got same output as you got and libass show back then

**Comment by behdad** [comment]
@khaledhosny please advise if there's anything you think we should fix.

**Comment by khaledhosny** [comment]
3. is non-issue, but 1. seems to be similar to #2647, except that the font has an empty `GDEF` table. 2. is a difference in mark zeroing, the font has an almost empty `GDEF` table and no glyphs are given mark class and I’m not sure what Uniscribe is using to decide it needs to zero the width of this glyph.

**Comment by khaledhosny** [comment]
> the font has an almost empty `GDEF` table and no glyphs are given mark class and I’m not sure what Uniscribe is using to decide it needs to zero the width of this glyph.

If I recover the code for zeroing marks by Unicode, and switch Arabic shaper to it, the issue is fixed and we match Uniscribe. But apparently Arabic shaper switched from zeroing by Unicode to zeroing by GDEF in f368ba4a9edec4e297616698097546e8e6c89e53 based on a request from me. Uniscribe behavior for the test from this commit did not change, so it seems it behavior is nuanced than zeroing by either Unicode or GDEF alone and follows some sort of hybrid approach. AFAICT, none of this is specified in OpenType ”spec” and one can only guess what Microsoft implementation is doing here.

**Comment by khaledhosny** [comment]
I tried something like the following and used it in the Arabic shaper:
```diff
diff --git a/src/hb-ot-shape.cc b/src/hb-ot-shape.cc
index 92831e022..539a96338 100644
--- a/src/hb-ot-shape.cc
+++ b/src/hb-ot-shape.cc
@@ -934,6 +934,23 @@ zero_mark_widths_by_gdef (hb_buffer_t *buffer, bool adjust_offsets)
     }
 }

+static inline void
+zero_mark_widths_by_gdef_or_unicode (hb_buffer_t *buffer, bool adjust_offsets)
+{
+  unsigned int count = buffer->len;
+  hb_glyph_info_t *info = buffer->info;
+  for (unsigned int i = 0; i < count; i++)
+    if (_hb_glyph_info_is_mark (&info[i]) ||
+        (!_hb_glyph_info_is_base_glyph (&info[i]) &&
+        !_hb_glyph_info_is_ligature (&info[i]) &&
+         _hb_glyph_info_get_general_category (&info[i]) == HB_UNICODE_GENERAL_CATEGORY_NON_SPACING_MARK))
+    {
+      if (adjust_offsets)
+       adjust_mark_offsets (&buffer->pos[i]);
+      zero_mark_width (&buffer->pos[i]);
+    }
+}
+
 static inline void
 hb_ot_position_default (const hb_ot_shape_context_t *c)
 {
@@ -1008,6 +1025,7 @@ hb_ot_position_complex (const hb_ot_shape_context_t *c)
       default:
       case HB_OT_SHAPE_ZERO_WIDTH_MARKS_NONE:
       case HB_OT_SHAPE_ZERO_WIDTH_MARKS_BY_GDEF_LATE:
+      case HB_OT_SHAPE_ZERO_WIDTH_MARKS_BY_GDEF_OR_UNICODE_LATE:
        break;
     }

@@ -1020,6 +1038,10 @@ hb_ot_position_complex (const hb_ot_shape_context_t *c)
        zero_mark_widths_by_gdef (c->buffer, adjust_offsets_when_zeroing);
        break;

+      case HB_OT_SHAPE_ZERO_WIDTH_MARKS_BY_GDEF_OR_UNICODE_LATE:
+       zero_mark_widths_by_gdef_or_unicode (c->buffer, adjust_offsets_when_zeroing);
+       break;
+
       default:
       case HB_OT_SHAPE_ZERO_WIDTH_MARKS_NONE:
       case HB_OT_SHAPE_ZERO_WIDTH_MARKS_BY_GDEF_EARLY:
```

It worked for the `Hacen Typographer Bold mod.ttf`, but when I tried to make a minimal test font, Uniscribe didn’t zero the mark width in the test font! After some fiddling, turns out Uniscribe will zero the width of glyphs with class 0 GDEF glyph class when they are the result of applying `mset` feature but otherwise will not zero the width.

**Comment by behdad** [comment]
> turns out Uniscribe will zero the width of glyphs with class 0 GDEF glyph class when they are the result of applying `mset` feature but otherwise will not zero the width.

Sigh...

I know at some point I switched to populating glyph classes from Unicode if the GDEF class was 0. But that caused problems.

How about this:

When substituting, etc, currently we carry the old glyph class forward to the new glyph if GDEF does not have glyph classes.

What if we carry forward the old glyph class if the new glyph's GDEF class is 0? We definitely shouldn't do that for Indic et al. That will make people scream. See:

  https://github.com/harfbuzz/harfbuzz/issues/347#issuecomment-267838368

**Comment by behdad** [comment]
Oh. How about this: same issue I referenced was about similar problem with Hebrew. The solution we settled on is to ignore GPOS and do fallback positioning if GPOS doesn't have Hebrew script. Feels like same for Arabic is appropriate?

**Comment by behdad** [comment]
This proof-of-concept fixes at least the first issue. Didn't try the second, but I guess it should fix both:
```
diff --git a/src/hb-ot-shape-complex-arabic.cc b/src/hb-ot-shape-complex-arabic.cc
index 1f244f940..e0a5c19b0 100644
--- a/src/hb-ot-shape-complex-arabic.cc
+++ b/src/hb-ot-shape-complex-arabic.cc
@@ -706,7 +706,7 @@ const hb_ot_complex_shaper_t _hb_ot_complex_shaper_arabic =
   nullptr, /* decompose */
   nullptr, /* compose */
   setup_masks_arabic,
-  HB_TAG_NONE, /* gpos_tag */
+  HB_TAG ('a','r','a','b'), /* gpos_tag. https://github.com/harfbuzz/harfbuzz/issues/2888#issuecomment-795871211 */
   reorder_marks_arabic,
   HB_OT_SHAPE_ZERO_WIDTH_MARKS_BY_GDEF_LATE,
   true, /* fallback_position */
```

This can't go in as-is, since our Arabic shaper handles both Arabic and Syriac. But if we agree on the logic, I find a way to make it work.

@jfkthame wdyt?

**Comment by jfkthame** [comment]
Makes sense as a solution, I guess, if you can figure out a tidy way to do it that covers both scripts handled by the shaper.

Though, is it possible a font might use the DFLT script tag and handle positioning there? Should that even be supported? It's too long since I looked at this stuff, don't remember.....

**Comment by behdad** [comment]
> Makes sense as a solution, I guess, if you can figure out a tidy way to do it that covers both scripts handled by the shaper.

I can definitely work that part out.  It used to be a callback: 48c513fec978819927535bc86b43be74315f746c

> Though, is it possible a font might use the DFLT script tag and handle positioning there? Should that even be supported? It's too long since I looked at this stuff, don't remember.....

Historically we allowed that. And we still allow for most shapers. Looks to me like Arabic & Hebrew were special-cased because they had non-OpenType shapers in Windows which was made up to work with pan-Unicode fonts like arial without (necessarily) adding the OT positioning for those scripts. So we are just tuning these two for that behavior.
